Importer tool with sample csv and defaults.json file

hitengajjar opened the pull request:
Use below command to get started 

```bash
cd panopticum
python3.6 ./manage.py importer --help
```
You may need time to understand `seeddata.csv` format. It will be improved over time.

gajjarhiten committed:
Importer tool to parse specific csv format. For correct working provide default.json as command argument. Use --help

perfguru87 commented on panopticum/management/commands/config/default.json:
```diff
@@ -0,0 +1,305 @@
+      "Hungary",
+      "South Africa",
+      "Sierra Leone",
+      "Kazakhstan"
```
M.b use alphabetical order?

hitengajjar replied:
Sure will do

hitengajjar replied:
I think we don't need to have countries as its not used at present. However even in case we keep it, we should use ISO3166 alpha3 codes.

At present, I considered to remove the list and take country alpha3 code directly from [iso3166 module](https://pypi.org/project/iso3166/).

den-gts replied:
may be we should move countires creation to [init.json](https://github.com/perfguru87/panopticum/blob/master/panopticum/fixtures/init.json) fixture and create it after db migration at deployment stage?

hitengajjar replied:
The reason why its not added to fixture is it overwrites the tables. Its better to use iso3166 module. But I really would like to drop this table as its not being used at the moment.

perfguru87 commented on panopticum/management/commands/config/default.json:
```diff
@@ -0,0 +1,305 @@
+   ],
+   "DataCenters":{
+      "EU2":{
+         "info":"",
```
We already have examples here: https://github.com/perfguru87/panopticum/raw/master/panopticum/static/images/panopticum-model.png

I'd propose to make them consistent:
Europe 1, USA 1

hitengajjar replied:
Got it, will do

perfguru87 commented on panopticum/management/commands/config/default.json:
```diff
@@ -0,0 +1,305 @@
+      "LinuxContainer": "Linux Container",
+      "LinuxOS": "Linux OS",
+      "MacOS": "Mac OS",
+      "plesk": "Plesk",
```
Plesk is not an environment, is it?

hitengajjar replied:
Its basically a web hosting platform. We can consider it as an environment unless you have different opinion.

perfguru87 replied:
Plesk is a Hosting Control Panel, it's an automation system, not an OS. You can't run an abstract process/server there and Plesk itself is run inside Linux/Windows

perfguru87 commented on panopticum/management/commands/config/default.json:
```diff
@@ -0,0 +1,305 @@
+         "description":"To be filled later by user."
+      },
+      "OSA":{
+         "order":13,
```
I think "location" should be more like "company datacenter", "centralized on-prem servers", "azure cloud", "end-user endpoint", etc.

But not an environment like "VmWare", "Android device", "iOS Device", etc.

Otherwise, it will be hard to distinguish the difference between "DeploymentEnvironments" and ""DeploymentLocation".

The pair of the Environment and Location defines the final target:
"Company datacenter" + "VmWare server"
"centralized on-prem server" + "Hyper-V"
"end-user endpoint" + "iOS Device"

hitengajjar replied:
[EDIT] Shouldn't we consider "Company datacenter" etc as an Environment and Location will be "VMWare", "Docker, "iOS" etc

Also I think "K8S" should be considered as Environment.

Please suggest your thoughts.

perfguru87 replied:
location is a particular point or place in physical space while environment is the surroundings of, and influences on, a particular item of interest (C)

VMWare, Hyper-V, etc are "Environments". They can be deployed in different locations: service provider datacenter, on-prem deployment "Locations".

"Environment" is a standard term now, e.g. "vSphere environment", "virtualized environment", "OS environment variables", etc.

perfguru87 commented on panopticum/management/commands/config/default.json:
```diff
@@ -0,0 +1,305 @@
+   },
+   "ProductFamilies":[
+      "Cloud Services",
+      "Native Applications"
```
"Cloud Services" and "Native Applications" sound like a 'component type'. ProductFamilies should be rather "My Cloud Product for Business", "My On-prem Product for Business", "My Desktop Product for Consumers"

https://github.com/perfguru87/panopticum/raw/master/panopticum/static/images/panopticum-model.png

hitengajjar replied:
Sure will align to panopticum-model.png

perfguru87 commented on panopticum/management/commands/importer.py:
```diff
@@ -0,0 +1,1085 @@
+    loggers = {'Logf'}
+
+    countries = \
+        {'Russia', 'United Kingdom', 'Japan', 'Singapore', 'Germany', 'Estonia'}
```
Why we have such a strange set of countries?

hitengajjar replied:
These were already part of one of the test DB instances, I carry forward. But we can define these with [country codes](https://en.wikipedia.org/wiki/ISO_3166) and country name.
It looks like we don't use the countries table, so we can drop this part altogether.

hitengajjar replied:
For now using iso3166 module and populating this table. So no need for seed data.
Kindly suggest if we should remove this table altogether.

perfguru87 commented on panopticum/management/commands/importer.py:
```diff
@@ -0,0 +1,1085 @@
+        {'Russia', 'United Kingdom', 'Japan', 'Singapore', 'Germany', 'Estonia'}
+
+    data_centers = \
+        {'EU2': {'info': '', 'grafana': 'https://grafana.company.com/display/EU2.html',
```
See my prev comments. And the question is still the same - why we need to hardcode it in both, script and config?

perfguru87 commented on panopticum/management/commands/importer.py:
```diff
@@ -0,0 +1,1085 @@
+# Provides all the utility functions either in static form or as part of instance methods.
+class Utility:
+    comp_categories = \
+        {'Category1': {'description': '', 'order': 1}}
```
Why we need to hardcode the data we have in config or csv files?

hitengajjar replied:
Initially it all started to have these structures with default values in code and then made it part of defaults.json. I kept them in code for users to see what structure each variable follows.

defaults.json will always overwrite the code defaults.

den-gts replied:
Let's initial data in [init.json](https://github.com/perfguru87/panopticum/blob/master/panopticum/fixtures/init.json)

perfguru87 commented on panopticum/management/commands/config/default.json:
```diff
@@ -0,0 +1,305 @@
+{
```
Why we need this file? Why can't we live with data only in csv file?

hitengajjar replied:
I began with CSV file which cannot have multiple sheets to represent seed data for all tables effectively. Hence came up with defaults.json.

It will converge to one xlsx file in the next version with different component details format. Till then we'll have 2 files with input data.

perfguru87 commented on panopticum/management/commands/importer.py:
```diff
@@ -0,0 +1,1085 @@
+    def usage():
+        print(r"Usage: ", file=sys.stderr)
+        print (r"Application to import CSV data into DB.")
+        print(r"# python ./manage.py seed-importer.py /path/to/seeddata.csv /path/to/defaults.json", file=sys.stderr)
```
Why we need both *.csv and *.json files?

hitengajjar replied:
I began with CSV file which cannot have multiple sheets to represent seed data for all tables effectively. Hence came up with defaults.json.

It will converge to one xlsx file in the next version with different component details format. Till then we'll have 2 files with input data.

perfguru87 commented on panopticum/management/commands/importer.py:
```diff
@@ -0,0 +1,1085 @@
+        # fn_name = 'Utility::seed_comp_privacies() -'
+        logging.info('starts')
+        for (name, values) in Utility.comp_dataprivacies.items():
+            dbobj = Utility.get_comp_privacy(name, values['order'])
```
You can probably add "raise_exception" parameter to get_comp_privacy() and optimize all the cases with:

obj = Utility.get_comp_privacy(..)
if obj is None:
     pass   # TODO: Exception

hitengajjar replied:
Got it, will add and let utility functions raise those exceptions.

perfguru87 commented on panopticum/management/commands/importer.py:
```diff
@@ -0,0 +1,1085 @@
+    @staticmethod
+    def seed_depl_envs():
+        # fn_name = 'Utility::seed_depl_envs() -'
+        logging.info('starts')
```
I see such logging.info('starts') pattern repeats everywhere and not informative, probably it's better to add good logging directly to alloc_obj_by_name()

hitengajjar replied:
[EDIT] Noted and will be removed in next commit.

perfguru87 commented on panopticum/management/commands/importer.py:
```diff
@@ -0,0 +1,1085 @@
+                for a in architects if architects is not None else []:
+                    comp_ver_obj.owner_architect.add(a)
+
+                # TODO: Check with AA how to get Program Manager list, escalation list.
```
?

hitengajjar replied:
Ignore this comment. The comment was added considering v1 csv format where we didn't have the information.

I will remove it in subsequent changes.

hitengajjar reviewed (commented):
I have provided my early comments. Feel free to reply back.
I will soon submit updated PR with necessary changes.

gajjarhiten committed:
Changes as per comments provided in PR 13 - https://github.com/perfguru87/panopticum/pull/13
changes to the location schema to consider shortname as it is widely used in the seedata template
Cosmetic changes to the final report for compversions update/created/failed counts
split framework and language seeding to separate functions
depl_locations code structure is optimized

gajjarhiten committed:
removing unncessary zip file

hitengajjar reviewed (commented):
Updated source. Request you to review the changes.

hitengajjar committed:
Merge branch 'master' into master

hitengajjar commented:
@perfguru87 I have incorporated multiple suggestions as mentioned in specific comments above. Some of the questions are answered.

perfguru87 reviewed (approved):
Let's remove all the data from *.json and *.py files and put everything to *.xlsx example
